plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts

```diff
@@ -0,0 +1,236 @@
+      return new SlackNotificationProcessor({
+        directWebhookUrl,
+        channelWebhookUrl,
+        ...options,
+      });
+    });
+  }
+
+  private constructor(options: {
+    auth: AuthService;
+    discovery: DiscoveryService;
+    logger: LoggerService;
+    directWebhookUrl: string;
+    channelWebhookUrl: string;
+    catalogClient?: CatalogClient;
+  }) {
+    const { auth, directWebhookUrl, channelWebhookUrl, logger } = options;
+    this.logger = logger;
+    this.directWebhookUrl = directWebhookUrl;
+    this.channelWebhookUrl = channelWebhookUrl;
+    this.auth = auth;
+    this.catalogClient =
+      options.catalogClient ??
+      new CatalogClient({
+        discoveryApi: options.discovery,
+      });
+  }
+  getName(): string {
+    return 'Slack';
+  }
+
+  async processOptions(
+    options: NotificationSendOptions,
+  ): Promise<NotificationSendOptions> {
+    if (options.recipients.type === 'broadcast') {
+      // Does not support broadcast notifications
+      return options;
+    }
+
+    const entityRefs = Array.isArray(options.recipients.entityRef)
+      ? options.recipients.entityRef
+      : [options.recipients.entityRef];
+
+    const outbound: Promise<void>[] = [];
+    await Promise.all(
+      entityRefs.map(async entityRef => {
+        const compoundEntityRef = parseEntityRef(entityRef);
+        if (compoundEntityRef.kind !== 'group') {
+          return;
+        }
+
+        let destination;
+        try {
+          destination = await this.getChannelFromGroup(entityRef);
+        } catch (error) {
+          this.logger.error(
+            `Failed to get Slack channel for group entity: ${error.message}`,
+          );
+          return;
+        }
+
+        if (!destination) {
+          this.logger.debug(
+            `No Slack channel found for group entity: ${entityRef}`,
+          );
+          return;
+        }
+
+        const payload = JSON.stringify({
+          ...options.payload,
+          destination: destination,
+        });
+
+        this.logger.debug(`Sending Slack channel notification: ${payload}`);
+        outbound.push(
+          sendSlackNotification(this.channelWebhookUrl, payload).catch(
+            error => {
+              this.logger.error(
+                `Failed to send Slack channel notification: ${error.message}`,
+              );
+            },
+          ),
+        );
+      }),
+    );
+
+    // this is a bit fire and forget, but I don't think we want to block further processing.
+    await Promise.all(outbound);
+
+    if (outbound.length > 0) {
+      // If we sent team notifications, we don't want to send individual ones after resolving the group
+      // This is how we prevent the other methods from sending individual notifications.
+      // Given the current logic, if a Group has no slack channel id annotation, we'll let it get expanded to individuals.
+      options.recipients.entityRef = [];
```

Comment:
I'm still trying to understand how we can provide the best configurability for notifications in general.

In this case I've chosen not to send user notifications if a group is the target entity but that may not always be what is desired.
It would be nice if the NotificationService allowed plugins to register types of notifications emitted. The notification preferences UI could then show all possible notification types and users could choose which NotificationProcessors (delivery methods) should be enabled for each notification type.

Reply:
You mean something like this: https://github.com/backstage/backstage/pull/23716 ?

Reply:
Yes.

That reminds me, my original draft had a configuration allowlist for "origins" so that we could easily control _which_ plugins were allowed to send slack notifications. For example you may want some plugins to send slack notifications, but others to limit themselves to web or other processors.

I removed it because origin is only available in the pre/post process methods and not in processOptions and I felt it would then provide an inconsistent experience.

Do you have any ideas on how we could include this? The whole thing may be moot once the notification settings are implemented, but this felt like a way to achieve at least some basic filtering in the meantime.